bin/runtest_common.ml

```diff
@@ -4,103 +4,227 @@ module Test_kind = struct
+            Source.Cram_test.path test |> Path.Source.to_string))
+      and+ ml_test_candidates =
+        Source_tree.Dir.filenames source_dir
+        |> Filename.Set.to_list
+        |> List.filter ~f:(String.is_suffix ~suffix:".ml")
+        |> Memo.List.filter ~f:(fun ml_file ->
+          classify_ml_test ~sctx ~dir:parent_dir ~ml_file >>| Result.is_ok)
+      and+ dir_candidates =
+        Source_tree.Dir.sub_dirs source_dir
+        |> String.Map.to_list
+        |> Memo.List.map ~f:(fun (_candidate, candidate_path) ->
+          Source_tree.Dir.sub_dir_as_t candidate_path
+          >>| Source_tree.Dir.path
+          >>| Path.Source.to_string)
+      in
+      List.concat [ cram_candidates; ml_test_candidates; dir_candidates ]
+      |> String.Set.of_list
+      |> String.Set.to_list
+  in
   User_error.raise
     ~hints:(User_message.did_you_mean (Path.Source.to_string path) ~candidates)
     [ Pp.textf "%S does not match any known test." (Path.Source.to_string path) ]
 ;;
 
-(* [disambiguate_test_name path] is a function that takes in a
-   directory [path] and classifies it as either a cram test or a directory to
-   run tests in. *)
-let disambiguate_test_name path =
-  match Path.Source.parent path with
-  | None -> Memo.return @@ Test_kind.Runtest (Path.source Path.Source.root)
-  | Some parent_dir ->
-    let open Memo.O in
-    let* cram_tests = cram_tests_of_dir parent_dir in
-    (match find_cram_test cram_tests path with
-     | Some test ->
-       (* If we find the cram test, then we request that is run. *)
-       Memo.return (Test_kind.Cram (Path.source parent_dir, test))
-     | None ->
-       (* If we don't find it, then we assume the user intended a directory for
-          @runtest to be used. *)
-       Source_tree.find_dir path
-       >>= (function
-        (* We need to make sure that this directory or file exists. *)
-        | Some _ -> Memo.return (Test_kind.Runtest (Path.source path))
-        | None -> explain_unsuccessful_search path ~parent_dir))
+(** Classifies [path] as a test by first checking its extension, then looking
+    it up via the appropriate finder. Returns a [Test_kind.t] for cram tests,
+    test executables, inline test libraries, or directories (for @runtest). *)
+let classify_test =
+  let open Memo.O in
+  let try_directory ~sctx ~parent_dir path =
+    Source_tree.find_dir path
+    >>= function
+    | Some _ -> Memo.return (Test_kind.Runtest (Path.source path))
+    | None -> explain_unsuccessful_search ~sctx path ~parent_dir
+  in
+  fun ~sctx path ->
+    match Path.Source.parent path with
+    | None -> Memo.return (Test_kind.Runtest (Path.source Path.Source.root))
+    | Some parent_dir ->
+      (* Rough classification by extension: .t files are likely cram tests,
+         implementation files (.ml, .re) are likely ML tests, and anything
+         else is assumed to be a directory. The actual validation happens
+         in the corresponding find_* functions below. *)
```

Comment:
Instead of relying on extensions, I think you should look retain the old logic for identifying cram tests. If something is found to be a cram test, then _it is_ a cram test. If that fails, you can try other possibilities.

You should also avoid extensions as much as posobile for identifying expect tests as well.

Reply:
Reverted this change